nano/node/socket.cpp

```diff
@@ -0,0 +1,334 @@
+			if (!write_in_progress)
+			{
+				this_l->write_queued_messages ();
+			}
+		});
+	}
+	else
+	{
+		start_timer ();
+		boost::asio::async_write (tcp_socket, boost::asio::buffer (buffer_a->data (), buffer_a->size ()),
+		boost::asio::bind_executor (strand,
+		[this_l, buffer_a, callback_a](boost::system::error_code const & ec, size_t size_a) {
+			if (auto node = this_l->node.lock ())
+			{
+				node->stats.add (nano::stat::type::traffic_tcp, nano::stat::dir::out, size_a);
+				this_l->stop_timer ();
+				if (callback_a)
+				{
+					callback_a (ec, size_a);
+				}
+			}
+		}));
+	}
+}
+
+void nano::socket::write_queued_messages ()
+{
+	std::weak_ptr<nano::socket> this_w (shared_from_this ());
+	auto msg (send_queue.front ());
+	start_timer ();
+	boost::asio::async_write (tcp_socket, boost::asio::buffer (msg.buffer->data (), msg.buffer->size ()),
+	boost::asio::bind_executor (strand,
+	[msg, this_w](boost::system::error_code ec, std::size_t size_a) {
+		if (auto this_l = this_w.lock ())
+		{
+			if (auto node = this_l->node.lock ())
+			{
+				node->stats.add (nano::stat::type::traffic_tcp, nano::stat::dir::out, size_a);
+			}
+			this_l->stop_timer ();
+
+			if (!this_l->closed)
+			{
+				if (msg.callback)
+				{
+					msg.callback (ec, size_a);
+				}
+
+				this_l->send_queue.pop_front ();
+				if (!ec && !this_l->send_queue.empty ())
+				{
+					this_l->write_queued_messages ();
+				}
+			}
+		}
+	}));
+}
+
+void nano::socket::start_timer ()
+{
+	if (auto node_l = node.lock ())
+	{
+		start_timer (node_l->config.tcp_io_timeout);
+	}
+}
+
+void nano::socket::start_timer (std::chrono::seconds deadline_a)
+{
+	std::weak_ptr<nano::socket> this_w (shared_from_this ());
+	boost::asio::post (strand, [this_w, deadline_a]() {
+		if (auto this_l = this_w.lock ())
+		{
+			this_l->deadline.expires_after (deadline_a);
+			this_l->deadline.async_wait (boost::asio::bind_executor (this_l->strand, [this_l](boost::system::error_code const & ec_a) {
+				if (!ec_a)
+				{
+					if (auto node_l = this_l->node.lock ())
+					{
+						this_l->timed_out = true;
+						this_l->close ();
+						if (node_l->config.logging.network_timeout_logging ())
+						{
+							node_l->logger.try_log (boost::str (boost::format ("Disconnecting from %1% due to timeout") % this_l->remote_endpoint ()));
+						}
+					}
+				}
+			}));
+		}
+	});
+}
+
+void nano::socket::stop_timer ()
+{
+	auto this_l (shared_from_this ());
+	boost::asio::dispatch (strand, [this_l]() {
+		if (this_l->max_idle_time && !this_l->closed)
+		{
+			this_l->start_timer (*this_l->max_idle_time);
+		}
+		else
+		{
+			this_l->deadline.cancel ();
+		}
+	});
+}
+
+bool nano::socket::has_timed_out () const
+{
+	return timed_out;
+}
+
+void nano::socket::set_max_idle_timeout (std::chrono::seconds max_idle_time_a)
+{
+	auto this_l (shared_from_this ());
+	boost::asio::dispatch (strand, [this_l, max_idle_time_a]() {
+		this_l->max_idle_time = max_idle_time_a;
+	});
+}
+
+// This must be called from a strand
```

Comment:
Is there a way to assert if we're inside a specific executor context?

Reply:
I had asserts on `running_in_this_thread`, but that didn't do what I thought (still asserts when called from a strand). I couldn't find a way, but can look again.